[wasm] Add support for AOT profiling, and add a `browser-profile` sample

tqiu8 opened the pull request:
Fix #44167

- Added profiling APIs for WASM
- Created new browser-profile sample for WASM 
  To get data.aotprofile: 
  `make get-aot-profile`

  To run with profile data:
  `make use-aot-profile`

- Add AotInterp (mixed mode) in order to use profile data. 
- Add Wasm config option to enable profiler.

tqiu8 committed:
Enable AOT Profiler and included option in wasm browser sample

tqiu8 committed:
Enable AOT Profiler and included option in wasm browser sample

lewing commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -1474,8 +1474,8 @@ var MonoSupportLib = {
 				MONO.mono_wasm_set_runtime_options (args.runtime_options);
-
-			if (args.aot_profiler_options)
+				
+			if (args.aot_profiler_options) {
```
This looks like it needs to be removed.

lewing replied:
The trailing brace is wrong.

lewing commented on src/mono/netcore/sample/wasm/browser/runtime.js:
```diff
@@ -9,7 +9,8 @@ var Module = {
             return fetch (asset, { credentials: 'same-origin' });
         }
-
+        MONO.mono_wasm_init_aot_profiler();
```
This shouldn't be needed.

tqiu8 replied:
Ok, I just need to define `config.aot_profiler_options` to have initialize the profiler.

lewing replied:
Yes, the build task should set it in the config

lewing commented on src/mono/netcore/sample/wasm/browser/runtime.js:
```diff
@@ -9,7 +9,8 @@ var Module = {
         MONO.mono_load_runtime_and_bcl_args (config);
     },
-};
+    runProfiler: true,
```
Is this used anywhere? Ah, you're using it to call StopProfile.  Typically you'd call that from managed, in the case of the sample it could just be at the end of main.

lewing commented on src/mono/wasm/runtime/library_mono.js:
```diff
@@ -1445,9 +1445,9 @@ var MonoSupportLib = {
 				options = {}
 			if (!('write_at' in options))
-				options.write_at = 'WebAssembly.Runtime::StopProfile';
+				options.write_at = 'Interop/Runtime::StopProfile';
```
If you change change this value you shouldn't need to load all the interop stuff.  You can point it at a function in the sample

tqiu8 replied:
To call `DumpAOTProfile` we would still have to load the interop stuff.

lewing replied:
The code is there already if the System.Private.Runtime.InteropServices.JavaScript assembly is being loaded (and DumpProfile hasn't been linked out)

lewing replied:
and by change the value I mean set it when generating the options.  The situation is kind of a mess because that assembly isn't public right now but we can deal with it differently until that changes.

tqiu8 replied:
Ah, got it.

lewing commented on src/mono/netcore/sample/wasm/browser/WasmSample.csproj:
```diff
@@ -38,15 +108,38 @@
 
   <ItemGroup>
     <Compile Include="Program.cs" />
+    <Compile Include="$(LibrariesProjectRoot)Common/src/Interop/Browser/Interop.Runtime.cs" Condition="'$(ProfileAOT)' == 'true'"/>
```
You won't need to add all this if change the write_at value in the options.

lewing commented on src/mono/netcore/sample/wasm/browser/Program.cs:
```diff
@@ -4,6 +4,8 @@
 using System;
 using System.Runtime.CompilerServices;
 
+[assembly: CLSCompliant(false)]
```
Remove this.

lewing commented on src/mono/netcore/sample/wasm/browser/index.html:
```diff
@@ -13,11 +13,32 @@ <h3 id="header">Wasm Browser Sample</h3>
     <script type='text/javascript'>
       var App = {
         init: function () {
+          Module.cwrap('mono_wasm_load_profiler_aot', 'void', ['string'])();
```
This shouldn't be needed.

lewing reviewed (changes requested):
Good start but needs some work. We should probably move this to a different sample separate it from the normal case.

tqiu8 committed:
PR feedback remove reference to Interop library, add profiler option to MonoAOTProfiler

tqiu8 committed:
PR feedback remove reference to Interop library, add profiler option to MonoAOTProfiler

radical commented on src/mono/netcore/sample/wasm/browser-profile/Makefile:
```diff
@@ -0,0 +1,29 @@
+clean:
+	rm -rf bin
+
+run:
```
You can use `dotnet-serve` instead, like https://github.com/dotnet/runtime/blob/master/src/mono/netcore/sample/wasm/browser/Makefile#L20-L26

mdh1418 replied:
In doing so, you can remove `server.py` and all other instances of it.

radical commented on tools-local/tasks/mobile.tasks/AotCompilerTask/MonoAOTCompiler.cs:
```diff
@@ -58,6 +58,10 @@ public class MonoAOTCompiler : Microsoft.Build.Utilities.Task
 
+    public string? AotProfilePath { get; set; }
+
+    public ITaskItem[]? Profilers { get; set; }
```
You can change the type to `string[]`. `ITaskItem` is useful if you want to access the metadata on the items, which is not the case here.

mdh1418 commented on src/mono/netcore/sample/wasm/browser-profile/WasmSample.csproj:
```diff
@@ -0,0 +1,127 @@
+             AssemblyFile="$(ArtifactsBinDir)WasmAppBuilder\Debug\$(NetCoreAppCurrent)\publish\WasmAppBuilder.dll"/>
+
+  <UsingTask TaskName="MonoAOTCompiler"
+             AssemblyFile="$(ArtifactsBinDir)MonoAOTCompiler\Debug\$(NetCoreAppCurrent)\MonoAOTCompiler.dll" />
```
I think `$(MonoAOTCompilerTasksAssemblyPath)` can be used here.

tqiu8 committed:
PR feedback: remove server.py and replace with dotnet-serve

tqiu8 committed:
PR feedback: remove server.py and replace with dotnet-serve

tqiu8 committed:
Merge branch 'master' into add-profiling-api

tqiu8 committed:
remove comments from WasmSample.csproj

tqiu8 committed:
remove comments from WasmSample.csproj

steveisok committed:
Add Wasm AOT steps into WasmApp.targets (#46081)

AOT can now be triggered through the libraries tests as well as the wasm samples.

Passing /p:RunAOTCompilation=true is all that is needed for the libraries test run. The property will be false (for now) by default.

steveisok committed:
Add Wasm AOT steps into WasmApp.targets (#46081)

AOT can now be triggered through the libraries tests as well as the wasm samples.

Passing /p:RunAOTCompilation=true is all that is needed for the libraries test run. The property will be false (for now) by default.

radical commented on src/mono/netcore/sample/wasm/browser-profile/runtime.js:
```diff
@@ -0,0 +1,15 @@
+        config.fetch_file_cb = function (asset) {
+            return fetch (asset, { credentials: 'same-origin' });
+        }
+        config.aot_profiler_options = {write_at:"Sample.Test::StopProfile", 
+                                        send_to: "System.Runtime.InteropServices.JavaScript.Runtime::DumpAotProfileData"}
```
formatting

radical commented on src/mono/netcore/sample/wasm/browser-profile/WasmSample.csproj:
```diff
@@ -0,0 +1,123 @@
+
+    <Error Condition="'$(RunAOTCompilation)' == 'true' and '$(EMSDK_PATH)' == ''" Text="The EMSDK_PATH environment variable should be set pointing to the emscripten SDK root dir when using AOT."/>
+    
+    <MonoAOTCompiler
```
The AOT stuff can be updated to shared targets in `WasmApp.targets`, like https://github.com/dotnet/runtime/blob/master/src/mono/netcore/sample/wasm/console/WasmSample.csproj .

tqiu8 committed:
fix profiler to run in non-AOT mode, style fixes

tqiu8 committed:
fix profiler to run in non-AOT mode, style fixes

tqiu8 committed:
clean up Makefile and driver.c

tqiu8 committed:
clean up Makefile and driver.c

tqiu8 committed:
Merge branch 'master' into add-profiling-api

tqiu8 committed:
Clean up Makefile add AotInterp mode to MonoAotMode

tqiu8 committed:
Clean up Makefile add AotInterp mode to MonoAotMode

tqiu8 committed:
formatting fixes

tqiu8 committed:
formatting fixes

tqiu8 committed:
Add space after 'fixed

tqiu8 committed:
Add space after 'fixed

radical commented:
I tried `get-aot-profile`, but I get errors in the browser:
![Screen Shot 2021-01-05 at 5 52 11 PM](https://user-images.githubusercontent.com/1472/103708670-021fa000-4f7f-11eb-8e57-ce3e2a034102.png)

This is with a clean build of the branch.

tqiu8 committed:
Enable BuildNative

tqiu8 committed:
Enable BuildNative

tqiu8 commented:
@radical Added a `/p:BuildNative=true` to the `build-aot-profile` target.

radical committed:
move the shared make logic to wasm.mk

radical committed:
move the shared make logic to wasm.mk

radical commented:
I would suggest moving the shared bits to a separate file - https://github.com/radical/runtime/commit/01946286c8553192c19184fc45e7fbe485b42fc8 .

radical commented on src/mono/netcore/sample/wasm/browser-profile/Makefile:
```diff
@@ -0,0 +1,33 @@
+	EMSDK_PATH=$(realpath $(TOP)/src/mono/wasm/emsdk) $(DOTNET) publish $(DOTNET_Q_ARGS) /p:EnableProfiler=true /p:WasmBuildNative=true /p:Configuration=$(CONFIG) /p:TargetArchitecture=wasm /p:TargetOS=Browser $(MSBUILD_ARGS) WasmSample.csproj
+
+run-aot-profiled:
+	cp $(HOME)/Downloads/data.aotprofile ./bin/$(CONFIG)/AppBundle
```
Instead of hardcoding the path, use a var here, and error out if that isn't given.

tqiu8 commented:
> I would suggest moving the shared bits to a separate file - [radical@0194628](https://github.com/radical/runtime/commit/01946286c8553192c19184fc45e7fbe485b42fc8) .

Could we open a separate PR for that after this one?

tqiu8 committed:
validate params add helix target

tqiu8 committed:
validate params add helix target

radical commented:
> > I would suggest moving the shared bits to a separate file - [radical@0194628](https://github.com/radical/runtime/commit/01946286c8553192c19184fc45e7fbe485b42fc8) .
> 
> Could we open a separate PR for that after this one?

why not just include it in this one?

radical commented:
> > > I would suggest moving the shared bits to a separate file - [radical@0194628](https://github.com/radical/runtime/commit/01946286c8553192c19184fc45e7fbe485b42fc8) .
> > 
> > 
> > Could we open a separate PR for that after this one?
> 
> why not just include it in this one?

Separate PR will be fine too.

radical commented on src/mono/netcore/sample/wasm/browser-profile/Makefile:
```diff
@@ -0,0 +1,36 @@
+	EMSDK_PATH=$(realpath $(TOP)/src/mono/wasm/emsdk) $(DOTNET) publish $(DOTNET_Q_ARGS) /p:EnableProfiler=true /p:WasmBuildNative=true /p:Configuration=$(CONFIG) /p:TargetArchitecture=wasm /p:TargetOS=Browser $(MSBUILD_ARGS) WasmSample.csproj
+
+run-aot-profiled:	
+ifeq ($(DOWNLOAD_DIR),)
```
Get the path to the profile data as input.

radical commented on src/mono/netcore/sample/wasm/browser-profile/Makefile:
```diff
@@ -0,0 +1,36 @@
+ifeq ($(DOWNLOAD_DIR),)
+	$(error DOWNLOAD_DIR is not set)
+endif
+	cp $(DOWNLOAD_DIR)/data.aotprofile bin/$(CONFIG)/AppBundle
```
Is this required? Or can we pass the original path to `/p:AOTProfilePath`?

tqiu8 replied:
Yes, we can do that. Just doing that to keep everything in one place.

tqiu8 committed:
Merge branch 'tq-add-profiling-api' of https://github.com/radical/runtime into radical-tq-add-profiling-api

tqiu8 committed:
Merge branch 'radical-tq-add-profiling-api' into add-profiling-api

radical commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -70,6 +70,11 @@ public static object GetGlobalObject(string? str = null)
             return Interop.Runtime.GetGlobalObject(str);
         }
 
+        public static void DumpAotProfileData (ref byte buf, int len, string s)
```
rename `s` to `extraArg`. That is what the profiler passes as the arg here. And same for the other `DumpAotProfileData` method.

radical commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -39,13 +39,19 @@
       Assemblies="@(AotInputAssemblies)"
       UseAotDataFile="false"
+      AotProfilePath="$(AOTProfilePath)"
+      Profilers="aot;"
```
Add a `@(AOTProfilers)` here, and you can add `aot` by default for that.

tqiu8 committed:
Merge branch 'master' into add-profiling-api

radical commented on src/tasks/WasmAppBuilder/WasmAppBuilder.cs:
```diff
@@ -217,6 +221,11 @@ public override bool Execute ()
                 if (source != null && source.ItemSpec != null)
                     config.RemoteSources.Add(source.ItemSpec);
         }
+        Console.WriteLine($"PROFILE {Profile}");
```
this line can be dropped

radical commented on src/tasks/WasmAppBuilder/WasmAppBuilder.cs:
```diff
@@ -25,6 +25,8 @@ public class WasmAppBuilder : Task
     [Required]
     public string[]? Assemblies { get; set; }
 
+    public bool Profile { get; set; }
```
Maybe `EnableProfiler` would be a better name?

radical commented on src/tasks/AotCompilerTask/MonoAOTCompiler.cs:
```diff
@@ -58,6 +58,16 @@ public class MonoAOTCompiler : Microsoft.Build.Utilities.Task
     public bool UseAotDataFile { get; set; } = true;
 
+    /// <summary>
+    /// Path to AOT profile file.
```
`File to use for profile-guided optimization`

radical commented on src/libraries/Common/src/Interop/Browser/Interop.Runtime.cs:
```diff
@@ -99,5 +100,26 @@ public static object GetGlobalObject(string? str = null)
+                var span = new ReadOnlySpan<byte>(p, len);
+                // Send it to JS
+                var module = (JSObject)Runtime.GetGlobalObject("Module");
+                module.SetObjectProperty("aot_profile_data", Uint8Array.From (span));
```
remove space before open paren

radical reviewed (changes requested):
Some more feedback! Thanks for addressing the earlier ones

tqiu8 committed:
Pr feedback

tqiu8 committed:
Pr feedback

tqiu8 committed:
fix index.html path

tqiu8 committed:
fix index.html path

radical commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -39,13 +39,19 @@
       <_WasmAssemblies Remove="@(_WasmAssemblies)" />
     </ItemGroup>
+    <PropertyGroup>
+      <AOTMode>LLVMOnly</AOTMode>
+      <AOTMode Condition="'$(BuildAOTProfiled)' == 'true'">AotInterp</AOTMode>
```
We should check to see if the property is already set. This would allow it to be overridden.

```suggestion
      <AOTMode Condition="'$(AOTMode)' == '' and '$(BuildAOTProfiled)' == 'true'">AotInterp</AOTMode>
      <AOTMode Condition="'$(AOTMode)' == ''">LLVMOnly</AOTMode>
```

radical commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -39,13 +39,19 @@
       OutputType="AsmOnly"
       Assemblies="@(AotInputAssemblies)"
       UseAotDataFile="false"
+      AotProfilePath="$(AOTProfilePath)"
```
Is it valid to have `AOTMode==AotInterp`(due to `BuildAOTProfiled==true`), and empty `AOTProfilePath`? If not, then we should check for that in the task.

tqiu8 replied:
Is it not possible to build in mixed mode without  a profile?

radical replied:
@vargaz can answer that question

radical commented on src/tasks/WasmAppBuilder/WasmAppBuilder.cs:
```diff
@@ -25,6 +25,8 @@ public class WasmAppBuilder : Task
     [Required]
     public string[]? Assemblies { get; set; }
 
+    public bool EnableProfile { get; set; }
```
```suggestion
    public bool EnableProfiler { get; set; }
```

radical commented on src/mono/netcore/sample/wasm/browser-profile/Wasm.BrowserProfile.Sample.csproj:
```diff
@@ -0,0 +1,81 @@
+        <WasmBuildDir>$(MSBuildProjectDirectory)\$(PublishDir)\</WasmBuildDir>
+        <WasmMainAssemblyPath>$(WasmBuildDir)$(AssemblyName).dll</WasmMainAssemblyPath>
+        <WasmMainJSPath>runtime.js</WasmMainJSPath>
+        <WasmMainAssemblyPath>$(WasmBuildDir)$(AssemblyName).dll</WasmMainAssemblyPath>
```
can be removed, as it is already set above

tqiu8 replied:
I'm going to remove the definition above. Since `WasmBuildDir` is still needed in defining `WasmMainAssemblyPath` etc which are required

radical commented on src/tasks/AotCompilerTask/MonoAOTCompiler.cs:
```diff
@@ -132,13 +142,19 @@ public override bool Execute()
+        }
+
         switch (Mode)
         {
             case "Normal": parsedAotMode = MonoAotMode.Normal; break;
             case "Full": parsedAotMode = MonoAotMode.Full; break;
             case "LLVMOnly": parsedAotMode = MonoAotMode.LLVMOnly; break;
+            case "AotInterp": parsedAotMode = MonoAotMode.AotInterp; break;
             default:
-                throw new ArgumentException($"'{nameof(Mode)}' must be one of: '{nameof(MonoAotMode.Normal)}', '{nameof(MonoAotMode.Full)}', '{nameof(MonoAotMode.LLVMOnly)}'. Received: '{Mode}'.", nameof(Mode));
+                throw new ArgumentException($"'{nameof(Mode)}' must be one of: '{nameof(MonoAotMode.Normal)}', '{nameof(MonoAotMode.Full)}', '{nameof(MonoAotMode.LLVMOnly)}', '{nameof(MonoAotMode.AotInterp)}'. Received: '{Mode}'.", nameof(Mode));
         }
```
```csharp
        if (!Enum.TryParse(Mode, true, out parsedAotMode))
        {
            Log.LogError($"Unknown Mode value: {Mode}. '{nameof(Mode)}' must be one of: {string.Join(',', Enum.GetNames(typeof(MonoAotMode)))}");
            return false;
        }
```

tqiu8 committed:
More PR feedback: add more checks to MonoAotCompiler task, change names

tqiu8 committed:
More PR feedback: add more checks to MonoAotCompiler task, change names

tqiu8 committed:
remove check for mixed mode and profile

tqiu8 committed:
remove check for mixed mode and profile

SamMonoRT commented:
@vargaz - please review.

radical reviewed (approved):
Looks good in general. I don't know much about the actual args being passed for aot, but other than that, LGTM! 👍

mdh1418 commented:
@tqiu8 If possible in this PR, it might be good to simplify the browser profile sample if @radical 's PR  https://github.com/dotnet/runtime/pull/46772 makes it in before this.

tqiu8 committed:
Merge branch 'master' into add-profiling-api

tqiu8 committed:
Update with new changes in wasm samples

tqiu8 committed:
Merge branch 'add-profiling-api' of github.com:tqiu8/runtime into add-profiling-api

vargaz commented:
The failure is because:
/bin/sh: 2: /tmp/tmpd4b55cd2d88f44fcb76e949b81b21acf.exec.cmd: source: not found
/__w/1/s/src/mono/wasm/build/WasmApp.targets(135,5): error MSB3073: The command "source /usr/local/emscripten/emsdk_env.sh && emcc --version | head -1 > emcc-version.txt" exited with code 127. [/__w/1/s/src/tests/FunctionalTests/wasm/AOT/browser/Wasm.Aot.Browser.Test.csproj]

tqiu8 committed:
remove duplicate msbuild task

radical commented:
Some makefile updates that can be added here, or in a separate PR: https://gist.github.com/radical/5fdb866f4c14795f67457d4e7634c1b1

tqiu8 committed:
Revert mono-cil-strip command

tqiu8 committed:
clean up merge

tqiu8 committed:
Merge branch 'master' into add-profiling-api

tqiu8 committed:
Merge branch 'master' into add-profiling-api

tqiu8 committed:
Merge branch 'add-profiling-api' of github.com:tqiu8/runtime into add-profiling-api

tqiu8 committed:
Restore comments

tqiu8 committed:
remove trailing spaces

tqiu8 committed:
add return false
